src/components/flyout/flyout.styles.ts

```diff
@@ -0,0 +1,282 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { css, keyframes } from '@emotion/react';
+import { EuiFlyoutPaddingSize } from './flyout';
+import { euiCanAnimate, euiBreakpoint, logicalCSS } from '../../global_styling';
+import { UseEuiTheme } from '../../services';
+import { euiShadowXLarge } from '../../themes/amsterdam/global_styling/mixins';
+import { transparentize } from '../../services/color';
+import { euiFormMaxWidth } from '../form/form.styles';
+
+const euiFlyout = keyframes`
+  0% {
+    opacity: 0;
+    transform: translateX(100%);
+  }
+  75% {
+    opacity: 1;
+    transform: translateX(0%);
+  }
+`;
+
+const euiFlyoutLeft = keyframes`
+  0% {
+    opacity: 0;
+    transform: translateX(-100%);
+  }
+  75% {
+    opacity: 1;
+    transform: translateX(0%);
+}
+`;
+
+export const euiFlyoutCloseButtonStyles = (euiThemeContext: UseEuiTheme) => {
+  const euiTheme = euiThemeContext.euiTheme;
+
+  return {
+    euiFlyout__closeButton: css`
+      background-color: ${transparentize(euiTheme.colors.emptyShade, 0.1)};
+      position: absolute;
+      ${logicalCSS('right', euiTheme.size.s)}
+      ${logicalCSS('top', euiTheme.size.s)}
+      z-index: 3;
+    `,
+    'closeButton--outside': css`
+      // match dropshadow
+      ${euiShadowXLarge(euiThemeContext)};
+      ${logicalCSS('right', 'auto')}
+      ${logicalCSS('left', 0)}
+      // Override the hover and focus transitions of buttons
+      animation: none !important;
+      ${euiBreakpoint(euiThemeContext, ['m', 'xl'])} {
+        transform: translateX(calc(-100% - ${euiTheme.size.l}));
+      }
+      ${euiBreakpoint(euiThemeContext, ['xs', 's'])} {
+        transform: translateX(calc(-100% - ${euiTheme.size.xs}));
+      }
+    `,
+    'closeButton--inside': css``,
+    'closeButton--outside-left': css`
+      ${logicalCSS('right', 0)}
+      ${logicalCSS('left', 'auto')}
+      
+      ${euiBreakpoint(euiThemeContext, ['m', 'xl'])} {
+        transform: translateX(calc(100% + ${euiTheme.size.l}));
+      }
+      ${euiBreakpoint(euiThemeContext, ['xs', 's'])} {
+        transform: translateX(calc(100% + ${euiTheme.size.xs}));
+      }
+    `,
+  };
+};
+
+export const euiFlyoutStyles = (
+  euiThemeContext: UseEuiTheme,
+  paddingSize: EuiFlyoutPaddingSize
+) => {
```

Comment:
see above comment, we shouldn't be passing paddingSize to this fn
```suggestion
export const euiFlyoutStyles = (euiThemeContext: UseEuiTheme) => {
```

Reply:
May I ask why? We're passing in `panelPadding` to `PopoverTitle` for `EuiPopover` and `paddingSize` for `EuiCard`. `EuiButton` and `EuiText` also pass in secondary parameters to their Emotion style functions.

Reply:
There's no need for it in this specific case. Each padding size can be generated statically by your `composeFlyoutPadding` util. Passing dynamic variables to Emotion causes Emotion to create new classNames on the fly which is going to be less performant and generate more cruft than static classNames (hence why we're still using, e.g. `style` to pass in custom colors rather than generating a new className for _n_ number of colors).

I [discussed this a while back](https://elastic.slack.com/archives/C7QC1JV6F/p1655753639684939?thread_ts=1655751225.728109&cid=C7QC1JV6F) and there are pros and cons to both (the primary pro being dev readability, and the primary con being Emotion performance).

> Just to close the loop on this from our discussion at sync today - this approach works well for EuiProgress which mostly handles conditional color CSS, but may not always work for all scenarios as Emotion will no longer statically generate all CSS together (conditional CSS is generated separately).
> tl;dr it depends!

To address the other examples you mentioned:

> EuiText also pass[es] in secondary parameters to their Emotion style functions

EuiText does not:
https://github.com/elastic/eui/blob/6d91a0981d44c43efb82862ff95f39e0e8f9443a/src/components/text/text.styles.ts#L216

I think you're looking at `euiScaleText` which is an internal util (basically akin to your `composeFlyoutPadding`). The architecture EuiText uses is exactly what I'm looking for in EuiFlyout, which is not passing a dynamic var from the React component to Emotion styles but using an internal composer to generate multiple outputs based on a parameter.

> EuiButton, EuiPopoverTitle, EuiCard

I'm looking at these examples closer and I'm absolutely not a fan of their usage of dynamic vars. the EuiButtonDisplay `minWidth` should 100% be using a `style` prop for dynamic min widths instead of Emotion. This is exactly the issue of generating too many classNames in Emotion that custom `color`s creates.

I also wasn't reviewer for this and it looks like a lot of Caroline's last PRs were kind of rushed in/not thoroughly reviewed, so it should probably be fixed. I'll look into opening up a follow-up PR. cc @thompsongl

Reply:
Yes let's take another pass and reevaluate with patterns that you've found helpful

Reply:
Sounds good. Thanks for the explanation